src/main/java/Core/Client.java

```diff
@@ -0,0 +1,46 @@
+package Core;
+
+import java.io.*;
+import java.net.Socket;
+
+public class Client {
+
+    private Socket socket;
+    private BufferedReader stdInReader;
+    private PrintStream stdOut;
+    private PrintStream dataSentToSocketPrinter;
+    private MessageListener messageListener;
+
+    public Client(InputStream stdIn, PrintStream stdOut, Socket socket) throws IOException {
```

Comment:
Your client over here has two responsibilities, that are both _pretty tricky on their own_:
1. Communication protocol, like dealing with `"bye"` and what to do when there are no more messages (`null`)
2. Starting a thread to be able to run multiple things.

I Highligh suggest you separate these two concerns.
Gently "mixing" in some concurrency at the bottom of a constructor will come at a cost.
Look at things like https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html and https://winterbe.com/posts/2015/04/07/java8-concurrency-tutorial-thread-executor-examples/